libkbfs/bare_root_metadata_v3.go

```diff
@@ -1006,50 +1001,129 @@ func (md *BareRootMetadataV3) SetRevision(revision MetadataRevision) {
+	ePubKey kbfscrypto.TLFEphemeralPublicKey,
+	ePrivKey kbfscrypto.TLFEphemeralPrivateKey,
+	tlfCryptKey kbfscrypto.TLFCryptKey) (UserDeviceKeyServerHalves, error) {
 	if md.TlfID().IsPublic() {
 		return nil, InvalidPublicTLFOperation{
-			md.TlfID(), "addKeyGenerationHelper", md.Version()}
+			md.TlfID(), "updateKeyBundles", md.Version()}
+	}
+
+	wkb, rkb, err := md.getTLFKeyBundles(extra)
+	if err != nil {
+		return nil, err
+	}
+
+	// No need to explicitly handle the reader rekey case.
+
+	var newWriterIndex int
+	if len(updatedWriterKeys) > 0 {
+		newWriterIndex = len(wkb.TLFEphemeralPublicKeys)
+	}
+	wServerHalves, err := wkb.Keys.fillInUserInfos(
+		crypto, newWriterIndex, updatedWriterKeys,
+		ePrivKey, tlfCryptKey)
+	if err != nil {
+		return nil, err
+	}
+	if len(wServerHalves) > 0 {
+		wkb.TLFEphemeralPublicKeys =
+			append(wkb.TLFEphemeralPublicKeys, ePubKey)
+	}
+
+	var newReaderIndex int
+	if len(updatedReaderKeys) > 0 {
+		newReaderIndex = len(rkb.TLFEphemeralPublicKeys)
+	}
+	rServerHalves, err := rkb.Keys.fillInUserInfos(
+		crypto, newReaderIndex, updatedReaderKeys,
+		ePrivKey, tlfCryptKey)
+	if err != nil {
+		return nil, err
+	}
+	if len(rServerHalves) > 0 {
+		rkb.TLFEphemeralPublicKeys =
+			append(rkb.TLFEphemeralPublicKeys, ePubKey)
+	}
+
+	return wServerHalves.mergeUsers(rServerHalves)
+}
+
+// AddKeyGeneration implements the MutableBareRootMetadata interface
+// for BareRootMetadataV3.
+func (md *BareRootMetadataV3) AddKeyGeneration(codec kbfscodec.Codec,
+	crypto cryptoPure, currExtra ExtraMetadata,
+	updatedWriterKeys, updatedReaderKeys UserDevicePublicKeys,
+	ePubKey kbfscrypto.TLFEphemeralPublicKey,
+	ePrivKey kbfscrypto.TLFEphemeralPrivateKey,
+	pubKey kbfscrypto.TLFPublicKey,
+	currCryptKey, nextCryptKey kbfscrypto.TLFCryptKey) (
+	nextExtra ExtraMetadata,
+	serverHalves UserDeviceKeyServerHalves, err error) {
+	if md.TlfID().IsPublic() {
+		return nil, nil, InvalidPublicTLFOperation{
+			md.TlfID(), "AddKeyGeneration", md.Version()}
 	}
+
+	if len(updatedWriterKeys) == 0 {
+		return nil, nil, errors.New(
+			"updatedWriterKeys unexpectedly non-empty")
+	}
+
 	if nextCryptKey == (kbfscrypto.TLFCryptKey{}) {
-		return nil, errors.New("Zero next crypt key")
+		return nil, nil, errors.New("Zero next crypt key")
 	}
+
 	latestKeyGen := md.LatestKeyGeneration()
 	var encryptedHistoricKeys EncryptedTLFCryptKeys
 	if currCryptKey == (kbfscrypto.TLFCryptKey{}) {
 		if latestKeyGen >= FirstValidKeyGen {
-			return nil, errors.Errorf(
+			return nil, nil, errors.Errorf(
 				"Zero current crypt key with latest key generation %d",
 				latestKeyGen)
 		}
 	} else {
+		currExtraV3, ok := currExtra.(*ExtraMetadataV3)
+		if !ok {
+			return nil, nil, errors.New("Invalid curr extra metadata")
+		}
+
+		existingWriterKeys := currExtraV3.wkb.Keys.toPublicKeys()
+		updatedWriterKeysRemoved := updatedWriterKeys.RemoveKeylessUsers()
+		if !existingWriterKeys.Equals(updatedWriterKeysRemoved) {
```

Comment:
Do you mean the last device for a writer? It did, which is why I had to add the `RemoveKeylessUsers` call.